### PR TITLE
New Threadpool

### PR DESCRIPTION
**Pull request by danmanj**
This adds new options to the set of existing threadpools. It also switches to using a threadpool for open-loop iteration.

### CODE REVIEW

File: distbench_engine.cc [hunk 1]
```diff
@@ -1091,15 +1093,13 @@ void DistBenchEngine::RunAction(ActionState* action_state) {
           copied_server_rpc_state->have_dedicated_thread = true;
           copied_server_rpc_state->SetFreeStateFunction(
               [=] { delete copied_server_rpc_state; });
-          RunRegisteredThread(
```
Comment (kanthicn1) [inline distbench_engine.cc:1094]: For my understanding, why did you remove this?

File: distbench_threadpool.cc [hunk 2]
```diff
@@ -74,51 +109,151 @@ SimpleThreadpool::~SimpleThreadpool() {
...
+    first_iteration = false;
+    if (work_mutex_.AwaitWithTimeout(working_conditions, timeout)) {
+      if (work_queue_.empty()) {
+        --idle_threads_;
+        --active_threads_;
+        work_mutex_.Unlock();
+        return;
+      }
+      // Work is available....
+      need_to_grow_threadpool = work_queue_.size() >= idle_threads_;
```
Comment (kanthicn1) [inline distbench_threadpool.cc:194]: So the algorithm is to grow threads as long as there are new work items in the queue?

File: distbench_threadpool.cc [hunk 3]
```diff
@@ -74,51 +109,151 @@ SimpleThreadpool::~SimpleThreadpool() {
...
+        --active_threads_;
+        work_mutex_.Unlock();
+        return;
+      } else {
+        work_mutex_.Unlock();
+        continue;
+      }
+    }
+    work_mutex_.Unlock();
+    if (need_to_grow_threadpool) {
```
Comment (kanthicn1) [inline distbench_threadpool.cc:220]: Why start a workerthread if you need to shrink threadpool?

File: distbench_threadpool.cc [hunk 4]
```diff
@@ -74,51 +109,151 @@ SimpleThreadpool::~SimpleThreadpool() {
...
+      // This thread is no longer idle, but.....
+      --idle_threads_;
+      // If we need to launch a new thread we want to account for that upcoming
+      // thread as soon as we drop the work_mutex_, to avoid spawning too
+      // many threads, so we must begin counting it here...
+      if (need_to_grow_threadpool) {
+        ++idle_threads_;
+      }
+    } else {
+      bool need_to_shrink_threadpool = idle_threads_ > max_idle_threads_;
```
Comment (kanthicn1) [inline distbench_threadpool.cc:207]: Is there no condition on total threads being more than max allowed? Is there such a thing as max allowed total threads (idle or not)?

**Review by kanthicn1 — APPROVED** [review]

File: distbench_engine.cc (hunk 1 above)
Comment (danmanj) [reply distbench_engine.cc:1094]: The old code created a new thread unconditionally, which has high overhead. The new code uses a threadpool interface to try to recycle threads.

File: distbench_threadpool.cc (hunk 2 above)
Comment (danmanj) [reply distbench_threadpool.cc:194]: This is intended to replace the code above which has no limits on the number of threads created, buttries to move the cost of launching threads into the threadpool as well, to increase the maximum ops per second that the calling code can achieve.

File: distbench_threadpool.cc (hunk 4 above)
Comment (danmanj) [reply distbench_threadpool.cc:207]: There is no limit, to be compatible with the old model of "just launch as many threads as you want"

File: distbench_threadpool.cc (hunk 3 above)
Comment (danmanj) [reply distbench_threadpool.cc:220]: This code needs to be cleaned up, so I've reverte this pull request to a draft, but in short the already existing thread can wake up after a period of inactivity and decide to exit if it is no longer needed.
